Xamarin.Forms.Platform.iOS/CollectionView/ObservableGroupedSource.cs

```diff
@@ -188,66 +181,64 @@ bool NotLoadedYet()
 		{
 			// If the UICollectionView hasn't actually been loaded, then calling InsertSections or DeleteSections is 
 			// going to crash or get in an unusable state; instead, ReloadData should be used
-			return !_collectionViewController.IsViewLoaded || _collectionViewController.View.Window == null;
+			return !_collectionViewController.IsViewLoaded || _collectionViewController.View?.Window == null;
 		}
 
-		async Task Add(NotifyCollectionChangedEventArgs args)
+		void Add(NotifyCollectionChangedEventArgs args)
 		{
-			if (ReloadRequired())
+			if (_collectionView.NumberOfSections() == 0 || _collectionView.VisibleCells.Length == 0)
 			{
-				await Reload();
+				Reload();
 				return;
 			}
 
-			var startIndex = args.NewStartingIndex > -1 ? args.NewStartingIndex : _groupSource.IndexOf(args.NewItems[0]);
+			var startIndex = args.NewStartingIndex > -1
+				? args.NewStartingIndex
+				: _originalItemsSource.IndexOf(args.NewItems[0]);
 			var count = args.NewItems.Count;
 
 			// Adding a group will change the section index for all subsequent groups, so the easiest thing to do
 			// is to reset all the group tracking to get it up-to-date
-			ResetGroupTracking();
+			ResetGroups();
 
 			// Queue up the updates to the UICollectionView
-			BatchUpdate(() => _collectionView.InsertSections(CreateIndexSetFrom(startIndex, count)));
+			_collectionView.InsertSections(CreateIndexSetFrom(startIndex, count));
```

Comment:
using BatchUpdate causes the concurrency problems in the first place. Just removing this here (and the other places) makes the updates a lot more reliable